### PR TITLE
added stream support to from_btl for issue #142

### PR DESCRIPTION
**Pull request by upsonp**

### CODE REVIEW

File: ctd/read.py [hunk 1]
```diff
@@ -61,6 +61,9 @@ def _open_compressed(fname):
 
 def _read_file(fname):
     """Read file contents."""
+    if isinstance(fname, StringIO):
+        return fname
```
Comment (ocefpaf) [inline ctd/read.py:65]: It is probably safer to add the fname.seek(0) and then `text = fname.copy()` (or something similar.)
We should also change the docstring to say "read from file or StringIO obj.

File: ctd/read.py [hunk 2]
```diff
@@ -261,7 +264,8 @@ def from_btl(fname):
 
     df["Statistic"] = df["Statistic"].str.replace(r"\(|\)", "")  # (avg) to avg
 
-    name = _basename(fname)[1]
+    if name is None:
+        name = _basename(fname)[1]
```
Comment (ocefpaf) [inline ctd/read.py:268]: This Forces folks using stream to always pass a name but that is optional, so the API here is kind of broken. (At least lacking documentation.)

Ideally we should check if it is stream and hardcode a default name for those cases.

For that to work we should move this check to `_read_file` and make it output both `f` and `name` there. That way we check for StringIO only once.

File: ctd/read.py (hunk 1 above)
Comment (upsonp) [reply ctd/read.py:65]: How about:

```python
    """Read file contents, or read from StringIO object."""
    if isinstance(fname, StringIO):
        fname.seek(0)
        text = fname.read()
        return StringIO(text)
```

Comment (ocefpaf) [reply ctd/read.py:65]: Looks good. Yeah, sorry for not checking I prototyped from my phone :-)
That looks good, not memory efficient but safe. We don't really need to be memory efficient b/c the files we are dealing with are never more than a few megabytes.

File: ctd/read.py (hunk 2 above)
Comment (upsonp) [reply ctd/read.py:268]: I hope I'm not over complicating this idea, but did you want `_read_file` to return a tuple of values, the first being the StringIO object and the second being the file/stream name?

Comment (ocefpaf) [reply ctd/read.py:270]: Yes. `_read_file` is an internal use function, so it is OK to "break it" and better than changing the call on `from_btl`. Does that make sense?

File: tests/test_read.py [hunk 3]
```diff
@@ -62,7 +62,7 @@ def btl_duplicate_header_name():
 def btl_as_stream():
```
Comment (ocefpaf) [inline tests/test_read.py:62]: This does not need to be a fixture b/c we won't be calling it multiple times in tests.
You can remove the  `@pytest.fixture` and rename it to `test_btl_as_stream` so pytest can execute it.

File: ctd/read.py (hunk 2 above)
Comment (upsonp) [reply ctd/read.py:270]: Here's an idea.

What if when reading the file, we grab the BTL "File Name" in the header section and use that to set the file name?

File: tests/test_read.py (hunk 3 above)
Comment (upsonp) [reply tests/test_read.py:62]: Sorry, I'm not use to unittesting specifically with pytest I was just emulating what other tests had done, I'll fix it up.

File: ctd/read.py (hunk 2 above)
Comment (upsonp) [reply ctd/read.py:270]: Sorry, what I mean is what if we got the file name in the _parse_seabird() method, if a name hasn't been set already.

File: tests/test_read.py (hunk 3 above)
Comment (ocefpaf) [reply tests/test_read.py:62]: Don't worry and no need to be sorry. The only reason I'm not sending commit directly to your branch is b/c I believe you want to do this. If not, I can take over. Your contributions and very welcomed and I'm trying to pay back by showing a bit how my messy tests work.

**Comment by upsonp** [comment]
@ocefpaf let me know what you think of this update.

I'm still mauling it over in my head. I'd prefer not changing default/existing behaviour so as not to accidentally break other people's code.

In this version, as discussed, I'm grabbing the name of the file from 'FileName' in the `_parse_seabird()` function. if the `metadata['name']` value isn't set there then I default to the old behaviour.

I'm making the assumption here that the names of all the files are the same ,e.g 185A007.BTL, 185A007.ROS, 185A007.hex, etc., so the stem of the name given by 'FileName' will be the same for all files

File: tests/test_read.py [hunk 4]
```diff
@@ -83,10 +85,18 @@ def test_btl_is_dataframe(btl):
...
+
+
+def test_btl_as_stringio(btl_as_stream):
+    assert isinstance(btl_as_stream, pd.DataFrame)
+    assert not btl_as_stream.empty
+
+
+def test_btl_as_stringio_without_name(btl_as_stream):
+    assert isinstance(btl_as_stream, pd.DataFrame)
+    assert not btl_as_stream.empty
```
Comment (ocefpaf) [inline tests/test_read.py:99]: `btl_as_stream` is from a stream and without a name. so you don't need the `test_btl_as_stringio` and `test_btl_as_stringio_without_name`, only one would suffice.

**Pull request merged**